File: Source/Fastore.Core/BTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Fastore.Core
{
    //Need to implement Delete, Get, Contains, Binary searching, etc.
    public class BTree<Key, Value> : IKeyValueTree<Key, Value>
    {
		public BTree(IComparer<Key> comparer = null, int fanout = 128, int leafSize = 128)
		{
			if (fanout < 2 || leafSize < 2)
				throw new ArgumentException("Minimum fan-out and leaf size is 2.");
			_fanout = fanout;
			_leafSize = leafSize;

			Comparer = comparer ?? Comparer<Key>.Default;
			_root = new Leaf(this);
		}

		public IComparer<Key> Comparer { get; private set; }

		private int _fanout = 10;
        public int Fanout
		{
			get { return _fanout; }
		}

        private int _leafSize = 100;
		public int LeafSize
		{
			get { return _leafSize; }
		}

		public event ValueMovedHandler<Key, Value> ValueMoved;
		
		private INode _root;

        public override string ToString()
        {
            return _root.ToString();
        }

		public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward)
		{
			return _root.Get(isForward);
		}

		public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start)
		{
			return _root.Get(isForward, start);
		}

		public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start, Optional<Key> end)
		{
			return _root.Get(isForward, start, end);
		}

		/// <summary> Attempts to insert a given key/value</summary>
		/// <param name="leaf"> The leaf in which the entry was located. </param>
		/// <returns> If the key already exists, nothing is changed and the existing value is returned. </returns>
        public Optional<Value> Insert(Key key, Value value, out IBTreeLeaf<Key,Value> leaf)
        {
            var result = _root.Insert(key, value, out leaf);
            if (result.Split != null)
                _root = new Branch(this, _root, result.Split.Right, result.Split.Key);
			return result.Found;
        }

        internal void DoValuesMoved(IBTreeLeaf<Key, Value> leaf)
        {
            if (ValueMoved != null)
				foreach (var entry in leaf.Get(true))
					ValueMoved(entry.Value, leaf);
        }

		class Branch : INode
		{
			public Branch(BTree<Key, Value> tree)
			{
				_tree = tree;
				_keys = new Key[tree.Fanout - 1];
				_children = new INode[tree.Fanout];
			}

			public Branch(BTree<Key, Value> tree, INode left, INode right, Key key)
				: this(tree)
			{
				_children[0] = left;
				_children[1] = right;
				_keys[0] = key;
				Count = 1;
			}

			private Key[] _keys;
			private INode[] _children;
			private BTree<Key, Value> _tree;

			/// <summary> Count is numbers of keys. Number of children is keys + 1. </summary>
			public int Count { get; private set; }

			public InsertResult Insert(Key key, Value value, out IBTreeLeaf<Key, Value> leaf)
			{
				var index = IndexOf(key);
				var result = _children[index].Insert(key, value, out leaf);

				// If child split, add the adjacent node
				if (result.Split != null)
					result.Split = InsertChild(index, result.Split.Key, result.Split.Right);
				return result;
			}

			private Split InsertChild(int index, Key key, INode child)
			{
				// If full, split
				if (Count == _tree.Fanout - 1)
				{
					int mid = (Count + 1) / 2;

					// Create new sibling node
					Branch node = new Branch(_tree);
					node.Count = Count - mid;
					Array.Copy(_keys, mid, node._keys, 0, node.Count);
					Array.Copy(_children, mid, node._children, 0, node.Count + 1);

					Count = mid - 1;

					Split result = new Split() { Key = _keys[mid - 1], Right = node };

					if (index <= Count)
						InternalInsertChild(index, key, child);
					else
						node.InternalInsertChild(index - (Count + 1), key, child);

					return result;
				}
				else
				{
					InternalInsertChild(index, key, child);
					return null;
				}
			}

			private void InternalInsertChild(int index, Key key, INode child)
			{
				int size = Count - index;
				Array.Copy(_keys, index, _keys, index + 1, size);
				Array.Copy(_children, index + 1, _children, index + 2, size);

				_keys[index] = key;
				_children[index + 1] = child;
				Count++;
			}

			private int IndexOf(Key key)
			{
				var result = Array.BinarySearch(_keys, 0, Count, key);
				if (result < 0)
				{
					var index = ~result;
					if (index > Count)
						return Count;
					else
						return index;
				}
				else
					return result;
			}

			public override string ToString()
			{
				var sb = new StringBuilder("[");
				for (int i = 0; i <= Count; i++)
					sb.Append(i.ToString() + ": " + _children[i].ToString());
				sb.Append("]");
				return sb.ToString();
			}

			public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward)
			{
				return Get(isForward, Optional<Key>.Null);
			}

			public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start)
			{
				return Get(isForward, start, Optional<Key>.Null);
			}

			public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start, Optional<Key> end)
			{
				var startIndex = start.HasValue ? IndexOf(start.Value) : 0;
				var endIndex = end.HasValue ? IndexOf(end.Value) : Count + 1;
				if (isForward)
				{
					for (int i = startIndex; i < endIndex; i++)
						foreach (var entry in _children[i].Get(isForward))
							yield return entry;
				}
				else
				{
					for (int i = endIndex - 1; i >= startIndex; i--)
						foreach (var entry in _children[i].Get(isForward))
							yield return entry;
				}
			}
		}

		class Leaf : INode, IBTreeLeaf<Key, Value>
		{
			private Value[] _values { get; set; }
			private Key[] _keys { get; set; }
			private int Count { get; set; }

			private BTree<Key, Value> _tree;

			public Leaf(BTree<Key, Value> parent)
			{
				_tree = parent;
				_keys = new Key[parent._leafSize];
				_values = new Value[parent._leafSize];
			}

			public InsertResult Insert(Key key, Value value, out IBTreeLeaf<Key, Value> leaf)
			{
				int pos = IndexOf(key);
				var result = new InsertResult();
				if (Count == _tree._leafSize)
				{
					var node = new Leaf(_tree);
					node.Count = (_tree._leafSize + 1) / 2;
					Count = Count - node.Count;

					Array.Copy(_keys, node.Count, node._keys, 0, node.Count);
					Array.Copy(_values, node.Count, node._values, 0, node.Count);

					if (pos < Count)
						result.Found = InternalInsert(key, value, pos, out leaf);
					else
						result.Found = node.InternalInsert(key, value, pos - Count, out leaf);

					_tree.DoValuesMoved(node);

					result.Split = new Split() { Key = node._keys[0], Right = node };
				}
				else
					result.Found = InternalInsert(key, value, pos, out leaf);

				return result;
			}

			public Optional<Value> InternalInsert(Key key, Value value, int index, out IBTreeLeaf<Key, Value> leaf)
			{
				leaf = this;
				if (index < Count && _tree.Comparer.Compare(_keys[index], key) == 0)
					return _values[index];
				else
				{
					Array.Copy(_keys, index, _keys, index + 1, Count - index);
					Array.Copy(_values, index, _values, index + 1, Count - index);
					_keys[index] = key;
					_values[index] = value;
					Count++;
					return Optional<Value>.Null;
				}
			}

			private int IndexOf(Key key)
			{
				var result = Array.BinarySearch(_keys, 0, Count, key, _tree.Comparer);
				if (result < 0)
				{
					var index = ~result;
					if (index > Count)
						return Count;
					else
						return index;
				}
				else
					return result;
			}

			public override string ToString()
			{
				var sb = new StringBuilder("{");
				var first = true;
				foreach (var entry in Get(true))
				{
					if (!first)
						sb.Append(",");
					else
						first = false;
					sb.Append(entry.Key);
					sb.Append(" : ");
					sb.Append(entry.Value);
				}
				sb.Append("}");
				return sb.ToString();
			}

			public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward)
			{
				return Get(isForward, Optional<Key>.Null);
			}

			public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start)
			{
				return Get(isForward, start, Optional<Key>.Null);
			}

			public IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start, Optional<Key> end)
			{
				var startIndex = start.HasValue ? IndexOf(start.Value) : 0;
				var endIndex = end.HasValue ? IndexOf(end.Value) : Count; 
				if (isForward)
				{
					for (int i = startIndex; i < endIndex; i++)
						yield return new KeyValuePair<Key, Value>(_keys[i], _values[i]);
				}
				else
				{
					for (int i = endIndex - 1; i >= startIndex; i--)
						yield return new KeyValuePair<Key, Value>(_keys[i], _values[i]);
				}
			}

			public Optional<Key> GetKey(Value value, IComparer<Value> comparer)
			{
				for (int i = 0; i < Count; i++)
				{
					if (comparer.Compare(_values[i], value) == 0)
						return _keys[i];
				}
				return Optional<Key>.Null;
			}
		}

		private interface INode
		{
			InsertResult Insert(Key key, Value value, out IBTreeLeaf<Key, Value> leaf);
			IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward);
			IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start);
			IEnumerable<KeyValuePair<Key, Value>> Get(bool isForward, Optional<Key> start, Optional<Key> end);
		}

		private struct InsertResult
		{
			public Optional<Value> Found;
			public Split Split;
		}

		private class Split
		{
			public Key Key;
			public INode Right;
		}
	}
}

File: Source/BtreeCPP/BTree.cpp
#include "BTree.h"
#include <string>
#include <sstream>

using namespace std;

//Tree

BTree::BTree(type keyType, type valueType, IObserver* observer)
{
	_keyType = keyType;
	_valueType = valueType;
	_branchCapacity = 128;
	_leafCapacity = 128;
	_observer = observer;
	_root = new Leaf(this);
}

BTree::~BTree()
{
	delete _root;
}

void* BTree::Insert(void* key, void* value, Leaf** leaf)
{
	InsertResult result = _root->Insert(key, value, leaf);
	if (result.split != NULL)
	{
		_root = new Branch(this, _root, result.split->right, result.split->key);
		delete result.split;
	}

	return result.found;
}

// Set the branch and leaf capacities; only affects nodes that are created after this is set
void BTree::setCapacity(int branchCapacity, int leafCapacity)
{
	_branchCapacity = branchCapacity;
	_leafCapacity = leafCapacity;
}

int BTree::getBranchCapacity()
{
	return _branchCapacity;
}

int BTree::getLeafCapacity()
{
	return _leafCapacity;
}

void BTree::DoValuesMoved(Leaf* leaf)
{
	if (_observer != NULL)
		_observer->ValuesMoved(leaf);
}

wstring BTree::ToString()
{
	return _root->ToString();
}

//Branch

Branch::Branch(BTree* tree)
{
	_tree = tree;
	_children = new INode*[_tree->_branchCapacity];
	_keys = new char[(_tree->_branchCapacity - 1) * _tree->_keyType.Size];
	Count = 0;
}

Branch::~Branch()
{
	delete[] _children;
	delete[] _keys;
}

Branch::Branch(BTree* tree, INode* left, INode* right, void* key)
{
	_tree = tree;
	_children = new INode*[_tree->_branchCapacity];
	_keys = new char[(_tree->_branchCapacity - 1) * _tree->_keyType.Size];
	_children[0] = left;
	_children[1] = right;
	memcpy(&_keys[0], key, _tree->_keyType.Size);
	Count = 1;
}

InsertResult Branch::Insert(void* key, void* value, Leaf** leaf)
{
	int index = IndexOf(key);
	InsertResult result = _children[index]->Insert(key, value, leaf);

	if (result.split != NULL)
	{
		Split* temp = result.split;
		result.split = InsertChild(index, result.split->key, result.split->right);
		delete temp;
	}
	return result;
}

Split* Branch::InsertChild(int index, void* key, INode* child)
{
	if (Count != _tree->_branchCapacity - 1)
	{
		InternalInsertChild(index, key, child);
		return NULL;
	}
	else
	{
		int mid = (Count + 1) / 2;
		Branch* node = new Branch(_tree);
		node->Count = Count - mid;
		memcpy(&node->_keys[0], &_keys[mid * _tree->_keyType.Size], node->Count * _tree->_keyType.Size);
		memcpy(&node->_children[0], &_children[mid], (node->Count + 1) * sizeof(INode*));
		
		Count = mid - 1;

		Split* split = new Split();
		split->key = &_keys[(mid - 1) * _tree->_keyType.Size];
		split->right = node;

		if (index <= Count)
			InternalInsertChild(index, key, child);
		else
			node->InternalInsertChild(index - (Count + 1), key, child);

		return split;
	}
}

void Branch::InternalInsertChild(int index, void* key, INode* child)
{
	int size = Count - index;
	if (Count != index)
	{
		memmove(&_keys[(index + 1) *_tree->_keyType.Size], &_keys[index * _tree->_keyType.Size],  size * _tree->_keyType.Size);
		memmove(&_children[index + 2], &_children[index + 1],  size * sizeof(INode*));
	}

	memcpy(&_keys[index * _tree->_keyType.Size], key, _tree->_keyType.Size);
	_children[index + 1] = child;
	Count++;
}

int Branch::IndexOf(void* key)
{	
    int lo = 0;
	int hi = Count - 1;
	int localIndex = 0;
	int result = -1;

	while (lo <= hi)
	{
		localIndex = (lo + hi) / 2;
        result = _tree->_keyType.Compare(key, &_keys[localIndex * _tree->_keyType.Size]);

		if (result == 0)
			break;
		else if (result < 0)
			hi = localIndex - 1;
		else
			lo = localIndex + 1;
	}

    if (result == 0)
        return localIndex;
    else
        return lo;
}

wstring Branch::ToString()
{
	wstringstream result;

	result << "\n[";
	bool first = true;
	for (int i = 0; i <= Count; i++)
	{
		if (!first)
			result << ",";
		else
			first = false;

		result << i << ": " << _children[i]->ToString();
	}
	result << "]";

	return result.str();
}

//Leaf

Leaf::Leaf(BTree* tree)
{
	_tree = tree;
	_keys = new char[tree->_leafCapacity * _tree->_keyType.Size];
	_values = new char[tree->_leafCapacity * _tree->_valueType.Size];
	Count = 0;
}

Leaf::~Leaf()
{
	delete[] _keys;
	delete[] _values;
}

InsertResult Leaf::Insert(void* key, void* value, Leaf** leaf)
{
	int index = IndexOf(key);
	InsertResult result;
	if (Count != _tree->_leafCapacity)
	{
		result.found = InternalInsert(index, key, value, leaf);
		result.split = NULL;
	}
	else
	{
		Leaf* node = new Leaf(_tree);
		if (index != Count)
		{
			node->Count = (_tree->_leafCapacity + 1) / 2;
			Count = Count - node->Count;

			memcpy(&node->_keys[0], &_keys[node->Count * _tree->_keyType.Size],  node->Count * _tree->_keyType.Size);
			memcpy(&node->_values[0], &_values[node->Count * _tree->_valueType.Size], node->Count * _tree->_valueType.Size);
		}

		if (index < Count)
			result.found = InternalInsert(index, key, value, leaf);
		else
			result.found = node->InternalInsert(index - Count, key, value, leaf);

		_tree->DoValuesMoved(node);

		Split* split = new Split();
		split->key = &node->_keys[0];
		split->right = node;

		result.split = split;

		return result;
	}

	return result;
}

void* Leaf::InternalInsert(int index, void* key, void* value, Leaf** leaf)
{
	*leaf = this;
	if (index >= Count || _tree->_keyType.Compare(&_keys[index * _tree->_keyType.Size], key) != 0)
	{
		if (Count != index)
		{
			memmove(&_keys[(index + 1) * _tree->_keyType.Size], &_keys[index], (Count - index) * _tree->_keyType.Size);
			memmove(&_values[(index + 1) * _tree->_valueType.Size], &_values[index], (Count - index) * _tree->_valueType.Size);
		}

		memcpy(&_keys[index * _tree->_keyType.Size], key, _tree->_keyType.Size);
		memcpy(&_values[index * _tree->_valueType.Size], value, _tree->_valueType.Size);
		Count++;
		return NULL;
	}
	else
		return &_values[index * _tree->_keyType.Size];
}

int Leaf::IndexOf(void* key)
{
	int lo = 0;
	int hi = Count - 1;
	int localIndex = 0;
	int result = -1;

	while (lo <= hi)
	{
		localIndex = (lo + hi) / 2;
        result = _tree->_keyType.Compare(key, &_keys[localIndex * _tree->_keyType.Size]);

		if (result == 0)
			break;
		else if (result < 0)
			hi = localIndex - 1;
		else
			lo = localIndex + 1;
	}

    if (result == 0)
        return localIndex;
    else
        return lo;
}

wstring Leaf::ToString()
{
	wstringstream result;
	result << "\n{";
	bool first = true;
	for(int i = 0; i < Count; i++)
	{
		if(!first)
			result << ",";
		else
			first = false;

		result << _tree->_keyType.ToString(&_keys[i * _tree->_keyType.Size]);
		result << ":";
		result << _tree->_valueType.ToString(&_values[i * _tree->_valueType.Size]);
	}
	result << "}";

	return result.str();
}

void* Leaf::GetKey(function<bool(void*)> predicate)
{
	for (int i = 0; i < Count; i++)
	{
		if (predicate(&_values[i * _tree->_valueType.Size]))
			return &_keys[i * _tree->_keyType.Size];
	}

	return NULL;
}
